src/tactic/declare.lean

```diff
@@ -0,0 +1,194 @@
+/-
+Copyright (c) 2021 Kyle Miller All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import tactic.core
+/-!
+# Canonical declarations
+
+This consists of two user-commands that allow you to define and use
+canonical declarations for certain concepts (for example, all of the
+`variables` to introduce algebraic objects such as a vector space).
+
+It like a parameterized version of `tactic.localized_cmd` (and the code
+is heavily inspired by it).
+
+See the tactic doc entry below for more information.
+-/
+
+open lean lean.parser interactive tactic native
+
+/-- Descriptor for a `def_declare`, constructed when building
+the attribute cache.
+It contains the number of parameters it expects, names for the
+metavariables given in the `def_declare` (for debugging), and
+a function that takes a list of that many parameters and produces
+the code to evaluate. -/
+meta structure canon_decl_def :=
+(nparams : ℕ)
+(default : list string)
+(mk_cmd : list string → tactic string)
+
+/-- The user attribute for canonical declarations. -/
+@[user_attribute]
+meta def canon_decl_attr : user_attribute (rb_lmap name canon_decl_def) unit := {
+  name := "_canon_decl_attr",
+  descr := "(interal) attribute for canonical declarations",
+  parser := failed,
+  cache_cfg := ⟨λ ns, (do dcls ← ns.mmap (λ n, mk_const n >>= eval_expr (name × canon_decl_def)),
+                          return $ rb_lmap.of_list dcls), []⟩ }
+
+/-- Get all commands for a given declaration and return them as a list -/
+meta def get_canon_decls (n : name) (nparams : ℕ) : tactic (list canon_decl_def) :=
+do m ← canon_decl_attr.get_cache,
+   match m.find n with
+   | [] := fail format!"canonical declaration {n} does not exist"
+   | decls := return $ decls.filter (λ decl, decl.nparams = nparams)
+   end
+
+/-- Get all commands for a given declaration and return them as a list -/
+meta def get_all_canon_decls (n : name) : tactic (list canon_decl_def) :=
+do m ← canon_decl_attr.get_cache,
+   match m.find n with
+   | [] := fail format!"canonical declaration {n} does not exist"
+   | decls := return $ decls
+   end
+
+/-- Execute all commands for the given type of canonical declaration. -/
+@[user_command] meta def declare_cmd (_ : parse $ tk "declare") : parser unit :=
+do do_print ← optional (tk "?"),
+   n ← ident,
+   args ← many ident,
+   let args' := args.map name.to_string,
+   decls ← get_canon_decls n args.length,
+   when (decls.length = 0) $
+     fail format!"no canonical declaration for {n} with {args.length} arguments",
+   decls.mmap' $ λ d, do
+     code ← d.mk_cmd args',
+     when do_print.is_some $ trace ("Try this: " ++ code),
+     emit_code_here code -- is it possible to print the code if this fails?
```

Comment:
```suggestion
     emit_code_here code <|> 
     (trace $ format!"The code\n  {code}\nis not a valid command.")
```

Reply:
What I meant to say in the comment is whether it's possible to both print the code and the original error that `emit_code_here` produces (I couldn't find anything that lets you capture the `exception` and use the error message yourself).  But I suppose it's reasonable to expect the user to try pasting in the code and seeing what the error was.

Reply:
Oh, I see. You can do that too. Define
```lean
meta def emit_code_here' (s : string) : lean.parser (option (option (unit → format))) := 
λ state, match emit_code_here s state with
| success a b := success none b
| exception a b c := success (some a) c
end
```
Then replace line 70 with
```suggestion
     some fmt ← emit_code_here' code | skip,
     let fmt := fmt.get_or_else (λ _, format!"unknown") (),
     interaction_monad.fail $ sformat!"The code\n  {code}\nfailed with error\n  {fmt}"
```

I'm not sure what the case is where a parser failure returns `none` instead of a thunked error message. The double option might not be necessary.